packages/fluentui/react-northstar/src/components/Tree/hooks/flattenTree.ts

```diff
@@ -0,0 +1,144 @@
+   * when tree item has no subtree, expanded is undefined
+   */
+  expanded?: boolean;
+
+  /**
+   * Also in TreeItemProps.
+   * parent id of the tree item. For the top level tree items, parent is undefined
+   */
+  parent?: string;
+
+  /**
+   * Also in TreeItemProps.
+   * Size of the tree/subtree that contains this item.
+   */
+  treeSize: number;
+
+  /** true if the tree item has subtree, indicating childrenIds are not undefined */
+  hasSubtree: boolean;
+
+  /** children ids of the tree item. For leaf tree item, childrenIds is undefined */
+  childrenIds?: string[];
+};
+
+export type FlatTree = {
+  [key: string]: BaseFlatTreeItem & { [key: string]: any };
+};
+
+export function flattenTree(
+  items: ObjectShorthandCollection<TreeItemProps>, // tree's props.items
+): FlatTree {
+  const result = {};
+  const flatten = (baseItems, level, parent?) => {
+    if (!baseItems) return;
+
+    if (parent) {
+      result[parent].childrenIds = [];
+    }
+
+    baseItems.forEach((item, indexAmongSiblings, arr) => {
+      const { id, items, expanded } = item;
+      if (parent) {
+        result[parent].childrenIds.push(id);
+      }
+
+      result[id] = {
+        index: indexAmongSiblings + 1, // Used for aria-posinset and it's 1-based.
+        level,
+        expanded: items?.length ? !!expanded : undefined,
+        parent: parent == null ? undefined : parent,
+        treeSize: arr.length,
+        hasSubtree: !!items?.length,
+      };
+
+      if (items?.length) {
+        flatten(items, level + 1, id);
+      }
+    });
+  };
+
+  flatten(items, 1);
+  return result;
+}
+
+export function getTobeRenderedItemsProps(
+  items: ObjectShorthandCollection<TreeItemProps>, // tree's props.items
+  flatTree: FlatTree,
+  treeRenderItemTitle?: ShorthandRenderFunction<TreeTitleProps>,
+): TreeItemProps[] {
+  if (!items) {
+    return [];
+  }
+
+  return items.reduce((prevProps: TreeItemProps[], currItem) => {
+    const {
+      id,
+      items,
+      title,
+      accessibility,
+      onFocusFirstChild,
+      onFocusParent,
+      onTitleClick,
+      onClick,
+      onSiblingsExpand,
+      renderItemTitle,
+      selectable,
+      selectionIndicator,
+    } = currItem;
+
+    if (!flatTree[id]) {
+      // should never happen
+      return prevProps;
+    }
+
+    const { expanded, parent, level, index, treeSize } = flatTree[id];
+
+    const treeItemProps = {
+      // user props:
+      id,
+      items,
+      key: id,
+      title,
+      ...(accessibility && { accessibility }), // to not override accessibility passed down from tree
+      onFocusFirstChild,
+      onFocusParent,
+      onTitleClick,
+      onClick,
+      onSiblingsExpand,
+      renderItemTitle: renderItemTitle ?? treeRenderItemTitle,
+      selectable,
+      selectionIndicator,
+      // props from flatTree:
+      expanded,
+      parent,
+      level,
+      index,
+      treeSize,
+    };
```

Comment:
I probably lost the idea there. I see two issues:
- tree is no way to pass additional props in this case
- I don't understand the value of flattening in this case as there we are doing it second time (?)

---

I suggest to modify `flattenTree` to include all necessary props instead

Reply:
I have to use items here, because flatTree does not keep order.

If I have tree item with id that can be parsed to int, for example '1' '2' ... They will always be in front of all other items in flatTree.

https://stackoverflow.com/questions/5525795/does-javascript-guarantee-object-property-order

Reply:
> tree is no way to pass additional props in this case

This can be done by iterate thru the returned `tobeRenderedItemsProps` again

> I don't understand the value of flattening in this case as there we are doing it second time (?)

Same reason as my previous comment. Flat tree does not keep order.

Another way could be, I remember all the root nodes of the tree, and generate `tobeRenderedItemsProps` based on their expanded props and childrenIds, and the ids of their children, and so on...